src/macros.rs

```diff
@@ -1,10 +1,305 @@
+///             // If there is no data in the message, you can use the unit type: `()`
+///             request_update $id $name: $body_type,
+///
+///             // The `response_update` keyword defines a response update.
+///             // You can have any amount of response updates inside a service definition.
+///             //
+///             // The $id is used as the service ID for the update and must be an i32.
+///             // The ID must be unique for all response updates in the service.
+///             //
+///             // The $name is the name of the update message.
+///             // It is used to generate function en type names.
+///             // It must be a valid Rust identifier and should be lower-case with underscores.
+///             //
+///             // The $body_type indicates the type of the message.
+///             // If there is no data in the message, you can use the unit type: `()`
+///             response_update $id $name: $body_type,
+///         }
+///
+///         // The `stream` keyword defines a stream message.
+///         // You can have any amount of stream definitions in an interface definition.
+///         //
+///         // The $id is used as the service ID of the stream message and must be an i32.
+///         // The ID must be unique for all streams in the interface.
+///         //
+///         // The $name is the name of the stream.
+///         // It is used to generate function en type names.
+///         // It must be a valid Rust identifier and should be lower-case with underscores.
+///         //
+///         // The $body_type indicates the type of the message.
+///         // If there is no data in the message, you can use the unit type: `()`
+///        stream $id $name: $body_type,
+///     }
+/// }
+/// ```
+///
+/// [`PeerHandle`]: crate::PeerHandle
+/// [`PeerWriteHandle`]: crate::PeerWriteHandle
+/// [`PeerReadHandle`]: crate::PeerReadHandle
 macro_rules! interface {
 	($($tokens:tt)*) => {
 		$crate::macros::interface_impl!{$crate; $($tokens)*}
 	}
 }
+
+/// Example module for the `interface!` macro.
+///
+/// You can compare the source the the gerated documentation to inspect the generated API.
+/// The most important generated types are [`supermarket::Client`] and [`supermarket::Server`].
+///
+/// [`supermarket::Client`]: interface_example::Client
+/// [`supermarket::Server`]: interface_example::Server
+/// ```
+///
+/// fizyr_rpc::interface! {
+///     /// RPC interface for the supermarket.
+///     interface supermarket {
+///         /// Greet the cashier.
+///         ///
+///         /// The cashier will reply with their own greeting.
+///         service 1 greet_cashier: String -> String,
+///
+///         /// Purchase tomatoes.
+///         ///
+///         /// The response of the cashier depends on the update messages exchanged.
+///         /// If you run away after they have sent the `price`, or if you pay with a nun-fungable token,
```

Comment:
I like tokens that don't become mushrooms, but I guess it is non-fungible

Reply:
Hmm, probably. Although either work equally as well as a money substitute.